server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -254,14 +254,27 @@ private void onFailure(int idx, String nodeId, Throwable t) {
         }
 
         private void finishHim() {
-            threadPool.executor(finalExecutor).execute(() -> newResponse(task, request, responses, listener));
+            if (isCancelled(task)) {
+                listener.onFailure(new TaskCancelledException("task cancelled"));
+                return;
+            }
+
+            final String executor = finalExecutor.equals(ThreadPool.Names.SAME) ? ThreadPool.Names.GENERIC : finalExecutor;
+            threadPool.executor(executor).execute(() -> newResponse(task, request, responses, listener));
         }
     }
 
-    class NodeTransportHandler implements TransportRequestHandler<NodeRequest> {
+    private boolean isCancelled(Task task) {
+        return task instanceof CancellableTask && ((CancellableTask) task).isCancelled();
+    }
 
+    class NodeTransportHandler implements TransportRequestHandler<NodeRequest> {
         @Override
         public void messageReceived(NodeRequest request, TransportChannel channel, Task task) throws Exception {
+            if (isCancelled(task)) {
+                throw new TaskCancelledException("task cancelled");
+            }
+
             channel.sendResponse(nodeOperation(request, task));
```

Comment:
Here I'm not sure if it's worth doing the check again and throwing after the `nodeOperation` have been executed when the task is cancelled. We could save some bandwidth as those responses would be discarded in the coordinating node anyway.

Reply:
I would say no. Throwing an exception could involve sending a few kB of stack trace over the wire so I don't think it makes much difference either way..